## Profiling (7 pts + 2 bonus pts)

Before we go any further and start looking at how vectorization makes your program faster, we need to talk about profiling. Profiling is the act of measuring performance of a program, either by timing it or by looking into memory access, depending on what is you are trying to measure.

(Follow the instructions here: https://jakevdp.github.io/PythonDataScienceHandbook/01.07-timing-and-profiling.html to setup the profilers)

# **Remember to save your file after generating all the required results. Then we can directly see your results.**

### Time

This is the most common profiler. In a python code you just import the time module and measure starting and ending time. For IPython we can call the %time %%time and %%timeit magic

In [14]:
%time?

# Question 1 (0.5 pts)
Run the following code and explain its output

In [15]:
%%time
total = 0
for i in range(1000):
    for j in range(1000):
        total += i * (-1) ** j

CPU times: user 1.38 s, sys: 4 ms, total: 1.39 s
Wall time: 1.4 s


The user time is 1.38 seconds of the 1.4 seconds of Wall time because most of our code is doing the function operations. It makes sense that the system time is low because we import no libraries and there is very little to memory to allocate.

# Question 2 (0.5)
There are two blocks of code below performing the same function on a given input, explain why the second sort is much faster

In [16]:
import random
L = [random.random() for i in range(100000)]
%time L.sort()

CPU times: user 140 ms, sys: 0 ns, total: 140 ms
Wall time: 164 ms


In [17]:
%time L.sort()

CPU times: user 16 ms, sys: 0 ns, total: 16 ms
Wall time: 17.1 ms


Similarly, to the previous part, the system has very little things to do and it is mostly the user program doing stuff. There is such a big descrepency beween the two codes because the the first block of code is randomized, meanwhile the second block is already sorted and lot less computations and comparisons are made.

# Question 3 (1 pts)
Use Python memory_profiler to profile your own code and explain the results

In [55]:
!pip3 install memory_profiler

Defaulting to user installation because normal site-packages is not writeable


In [56]:
%reload_ext memory_profiler

In [62]:
%%file something.py
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i ) for j in range(N)]
        total += sum(L)
        del L
    return total

Overwriting something.py


In [63]:
from something import sum_of_lists
%mprun -f sum_of_lists sum_of_lists(10000)

Our code gets the sum all integers from 1 to N. Our profile pager shows that we allocated ~74 MiB for our program. We had a spike in the initial definition of the function and we had a large negative decrement in line 4 as we updated our total time.The negative deallocation may be due to how fast we deallocated our resources,deallocating the resource before the profiler logs a use of that resource.

# Question 4 (7 pts)
Run the following codes to measure execution time, memory usage and answer the following questions.
Note: Make sure to install any missing Python packages

1. This code snippet defines and runs a simple Python function hello() that prints 'hello world!'. It also employs the memory_profiler module to profile the memory usage of the hello() function with a specified precision.

In [72]:
%%file helloworld.py
from memory_profiler import profile

@profile(precision=4)
def hello():
	print("hello world!") 

hello()

Overwriting helloworld.py


In [74]:
%run -t helloworld.py

hello world!
Filename: /home/aishyash/Downloads/helloworld.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     3  73.6836 MiB  73.6836 MiB           1   @profile(precision=4)
     4                                         def hello():
     5  73.6836 MiB   0.0000 MiB           1   	print("hello world!") 



IPython CPU timings (estimated):
  User   :       0.01 s.
  System :       0.00 s.
Wall time:       0.01 s.


2. This code snippet demonstrates memory profiling for a Python function my_func() that creates, manipulates, and deletes large lists, showcasing how memory usage changes with these operations

In [75]:
%%file expressions.py
from memory_profiler import profile
@profile(precision=4)
def my_func():
    a = [1] * (10 ** 6)
    b = [2] * (2 * 10 ** 7)
    del b
    return a
my_func()

Writing expressions.py


In [77]:
%run -t expressions.py

Filename: /home/aishyash/Downloads/expressions.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     2  73.8359 MiB  73.8359 MiB           1   @profile(precision=4)
     3                                         def my_func():
     4  73.8359 MiB   0.0000 MiB           1       a = [1] * (10 ** 6)
     5 226.2266 MiB 152.3906 MiB           1       b = [2] * (2 * 10 ** 7)
     6  73.8359 MiB -152.3906 MiB           1       del b
     7  73.8359 MiB   0.0000 MiB           1       return a



IPython CPU timings (estimated):
  User   :       0.14 s.
  System :       0.12 s.
Wall time:       0.27 s.


3. This code snippet profiles memory usage of the function math_funcs(), which demonstrates the application of logarithmic, cosine, and reciprocal functions from the NumPy library on an array of numbers, and prints the results for each operation.

In [78]:
%%file math_funcs.py
from memory_profiler import profile
import math
import numpy as np

@profile(precision=4)
def math_funcs():
	inp_arr = [10, 20, 30, 40, 50] 
	print ("Array input elements:\n", inp_arr) 

	res_arr = np.log(inp_arr) 
	print ("Applying log function:\n", res_arr)

	res_arr2 = np.cos(inp_arr) 
	print ("Applying cos function:\n", res_arr2)

	res_arr3 = np.reciprocal(inp_arr) 
	print ("Applying reciprocal function:\n", res_arr3)


math_funcs()

Writing math_funcs.py


In [79]:
%run -t math_funcs.py

Array input elements:
 [10, 20, 30, 40, 50]
Applying log function:
 [ 2.30258509  2.99573227  3.40119738  3.68887945  3.91202301]
Applying cos function:
 [-0.83907153  0.40808206  0.15425145 -0.66693806  0.96496603]
Applying reciprocal function:
 [0 0 0 0 0]
Filename: /home/aishyash/Downloads/math_funcs.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     5  79.9609 MiB  79.9609 MiB           1   @profile(precision=4)
     6                                         def math_funcs():
     7  79.9609 MiB   0.0000 MiB           1   	inp_arr = [10, 20, 30, 40, 50] 
     8  79.9609 MiB   0.0000 MiB           1   	print ("Array input elements:\n", inp_arr) 
     9                                         
    10  79.9609 MiB   0.0000 MiB           1   	res_arr = np.log(inp_arr) 
    11  79.9609 MiB   0.0000 MiB           1   	print ("Applying log function:\n", res_arr)
    12                                         
    13  79.9609 MiB   0.0000 MiB           1   	res_arr2 = n

4. This code snippet, using memory profiling, demonstrates a nested loop in Python where it iterates through combinations of adjectives and fruit names, printing each pair.

In [80]:
%%file loops.py
from memory_profiler import profile
import numpy as np
import ctypes
import math
import time

@profile(precision=4)
def my_loops():
	adj = ["red", "big", "tasty"]
	fruits = ["apple", "banana", "cherry"]

	for x in adj:
 		 for y in fruits:
   			 print(x, y)


my_loops()

Writing loops.py


In [82]:
%run -t loops.py

red apple
red banana
red cherry
big apple
big banana
big cherry
tasty apple
tasty banana
tasty cherry
Filename: /home/aishyash/Downloads/loops.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     7  80.2500 MiB  80.2500 MiB           1   @profile(precision=4)
     8                                         def my_loops():
     9  80.2500 MiB   0.0000 MiB           1   	adj = ["red", "big", "tasty"]
    10  80.2500 MiB   0.0000 MiB           1   	fruits = ["apple", "banana", "cherry"]
    11                                         
    12  80.2500 MiB   0.0000 MiB           4   	for x in adj:
    13  80.2500 MiB   0.0000 MiB          12    		 for y in fruits:
    14  80.2500 MiB   0.0000 MiB           9      			 print(x, y)



IPython CPU timings (estimated):
  User   :       0.04 s.
  System :       0.00 s.
Wall time:       0.05 s.


## Question 4.1 (1.5 pts)
Modify each of the above function to capture their execution time (Both CPU and Wall). You can modify the code directly, if required.

In [ ]:
# You can modify the code in-place or re-write the code here --- did it in original place

## Question 4.2 (1.5 pts)

What patterns did you notice between each of the above function with respect to latency, memory usage and code ?


Often, the Wall time was close to the sum of User and System time. Additionally, the higher the complexity of the function, the longer the system took to calculate the values (especially with progran 2). Snippets 1 and 4 were fairly simple, so it made sense that they occured in less than 0.1 seconds. The strange function was snippet 3 in which our Wall time was significantly higher than the sum of User and System times. This may be due to waiting by system during runtime.

## Question 4.3 (2 pts)
Using %time magic command, we can trace overall code execution time. Sometimes, you might have to get deeper insights to identify performance bottlenecks. Write your own code and profile execution time line by line.

In [110]:
%%file compute.py


from memory_profiler import profile

@profile(precision=4)
def compute(N):
    result = 0
    for i in range(N):
        L = i + i +1
        result += L
    return result

compute(100)

'''%%file file_for_some_function.py
%%timeit
def some_function(N):
    total = 0
    for i in range(N):
        L = i + i + 1
        total += L
        del L
    return total'''

Overwriting compute.py


In [111]:
%run -t compute.py

Filename: /home/aishyash/Downloads/compute.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     5  80.3398 MiB  80.3398 MiB           1   @profile(precision=4)
     6                                         def compute(N):
     7  80.3398 MiB   0.0000 MiB           1       result = 0
     8  80.3398 MiB   0.0000 MiB         101       for i in range(N):
     9  80.3398 MiB   0.0000 MiB         100           L = i + i +1
    10  80.3398 MiB   0.0000 MiB         100           result += L
    11  80.3398 MiB   0.0000 MiB           1       return result



IPython CPU timings (estimated):
  User   :       0.09 s.
  System :       0.02 s.
Wall time:       0.12 s.


'from file_for_some_function import some_function\n%mprun -f some_function some_function(100)'

## (Bonus) Question 4.4 (2 pts)
Memory usage of a program can also be reported as a function of time. Profile memory of any of the above code as a function of time.
Submit your profile results and a plot of the results (Mem used vs Time).

In [ ]:
# Your code goes here

Plot goes here